File: ref/algorithmic-trading/apps/app/app/src/test/java/com/trading/app/domain/order/OrderStateMachineTest.kt
package com.trading.app.domain.order

import com.trading.app.domain.model.OrderStatus
import org.junit.Assert.*
import org.junit.Test

class OrderStateMachineTest {

    @Test
    fun `initial status is CREATED`() {
        val sm = OrderStateMachine()
        assertEquals(OrderStatus.CREATED, sm.status)
        assertFalse(sm.isTerminal)
    }

    @Test
    fun `CREATED can transition to SUBMITTED`() {
        val sm = OrderStateMachine()
        assertTrue(sm.canTransition(OrderStatus.SUBMITTED))
        sm.transition(OrderStatus.SUBMITTED)
        assertEquals(OrderStatus.SUBMITTED, sm.status)
    }

    @Test
    fun `CREATED cannot transition to FILLED`() {
        val sm = OrderStateMachine()
        assertFalse(sm.canTransition(OrderStatus.FILLED))
    }

    @Test(expected = IllegalStateException::class)
    fun `invalid transition throws`() {
        val sm = OrderStateMachine()
        sm.transition(OrderStatus.FILLED)
    }

    @Test
    fun `full happy path - CREATED to FILLED`() {
        val sm = OrderStateMachine()
        sm.transition(OrderStatus.SUBMITTED)
        sm.transition(OrderStatus.PENDING)
        sm.transition(OrderStatus.FILLED)
        assertEquals(OrderStatus.FILLED, sm.status)
        assertTrue(sm.isTerminal)
    }

    @Test
    fun `cancellation path`() {
        val sm = OrderStateMachine()
        sm.transition(OrderStatus.SUBMITTED)
        sm.transition(OrderStatus.PENDING)
        sm.transition(OrderStatus.CANCEL_REQUESTED)
        sm.transition(OrderStatus.CANCELLED)
        assertEquals(OrderStatus.CANCELLED, sm.status)
        assertTrue(sm.isTerminal)
    }

    @Test
    fun `CANCEL_REQUESTED can transition to FILLED (race condition)`() {
        val sm = OrderStateMachine()
        sm.transition(OrderStatus.SUBMITTED)
        sm.transition(OrderStatus.PENDING)
        sm.transition(OrderStatus.CANCEL_REQUESTED)
        assertTrue(sm.canTransition(OrderStatus.FILLED))
    }

    @Test
    fun `partial fill path`() {
        val sm = OrderStateMachine()
        sm.transition(OrderStatus.SUBMITTED)
        sm.transition(OrderStatus.PENDING)
        sm.transition(OrderStatus.PARTIAL_FILLED)
        sm.transition(OrderStatus.FILLED)
        assertTrue(sm.isTerminal)
    }

    @Test
    fun `REJECTED is terminal`() {
        val sm = OrderStateMachine()
        sm.transition(OrderStatus.SUBMITTED)
        sm.transition(OrderStatus.REJECTED)
        assertTrue(sm.isTerminal)
        assertTrue(sm.allowedTransitions().isEmpty())
    }

    @Test
    fun `onTransition callback is called`() {
        val transitions = mutableListOf<Pair<OrderStatus, OrderStatus>>()
        val sm = OrderStateMachine(onTransition = { from, to -> transitions.add(from to to) })
        sm.transition(OrderStatus.SUBMITTED)
        sm.transition(OrderStatus.PENDING)

        assertEquals(2, transitions.size)
        assertEquals(OrderStatus.CREATED to OrderStatus.SUBMITTED, transitions[0])
        assertEquals(OrderStatus.SUBMITTED to OrderStatus.PENDING, transitions[1])
    }
}

File: ref/algorithmic-trading/apps/engine/src/state/order-state-machine.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { OrderStateMachine } from './order-state-machine.js';
import { OrderStatus } from '@trading/shared/types';

describe('OrderStateMachine', () => {
  it('starts with CREATED status', () => {
    const sm = new OrderStateMachine();
    assert.equal(sm.status, OrderStatus.CREATED);
  });

  it('transitions CREATED → SUBMITTED → PENDING → FILLED', () => {
    const sm = new OrderStateMachine();
    sm.transition(OrderStatus.SUBMITTED);
    assert.equal(sm.status, OrderStatus.SUBMITTED);

    sm.transition(OrderStatus.PENDING);
    assert.equal(sm.status, OrderStatus.PENDING);

    sm.transition(OrderStatus.FILLED);
    assert.equal(sm.status, OrderStatus.FILLED);
    assert.equal(sm.isTerminal, true);
  });

  it('transitions PENDING → PARTIAL_FILLED → FILLED', () => {
    const sm = new OrderStateMachine(OrderStatus.PENDING);
    sm.transition(OrderStatus.PARTIAL_FILLED);
    assert.equal(sm.status, OrderStatus.PARTIAL_FILLED);

    sm.transition(OrderStatus.FILLED);
    assert.equal(sm.status, OrderStatus.FILLED);
    assert.equal(sm.isTerminal, true);
  });

  it('transitions PENDING → CANCEL_REQUESTED → CANCELLED', () => {
    const sm = new OrderStateMachine(OrderStatus.PENDING);
    sm.transition(OrderStatus.CANCEL_REQUESTED);
    sm.transition(OrderStatus.CANCELLED);
    assert.equal(sm.status, OrderStatus.CANCELLED);
    assert.equal(sm.isTerminal, true);
  });

  it('transitions SUBMITTED → REJECTED', () => {
    const sm = new OrderStateMachine(OrderStatus.SUBMITTED);
    sm.transition(OrderStatus.REJECTED);
    assert.equal(sm.status, OrderStatus.REJECTED);
    assert.equal(sm.isTerminal, true);
  });

  it('allows CANCEL_REQUESTED → FILLED (race condition)', () => {
    const sm = new OrderStateMachine(OrderStatus.CANCEL_REQUESTED);
    sm.transition(OrderStatus.FILLED);
    assert.equal(sm.status, OrderStatus.FILLED);
  });

  it('throws on invalid transition', () => {
    const sm = new OrderStateMachine();
    assert.throws(
      () => sm.transition(OrderStatus.FILLED),
      /Invalid order state transition: CREATED → FILLED/
    );
  });

  it('throws on transition from terminal state', () => {
    const sm = new OrderStateMachine(OrderStatus.FILLED);
    assert.throws(
      () => sm.transition(OrderStatus.CANCELLED),
      /Invalid order state transition/
    );
  });

  it('canTransition returns correct boolean', () => {
    const sm = new OrderStateMachine();
    assert.equal(sm.canTransition(OrderStatus.SUBMITTED), true);
    assert.equal(sm.canTransition(OrderStatus.FILLED), false);
  });

  it('allowedTransitions returns valid next states', () => {
    const sm = new OrderStateMachine(OrderStatus.PENDING);
    const allowed = sm.allowedTransitions();
    assert.deepEqual(allowed, [
      OrderStatus.FILLED,
      OrderStatus.PARTIAL_FILLED,
      OrderStatus.REJECTED,
      OrderStatus.CANCEL_REQUESTED
    ]);
  });

  it('transitions PARTIAL_FILLED → CANCEL_REQUESTED → CANCELLED', () => {
    const sm = new OrderStateMachine(OrderStatus.PARTIAL_FILLED);
    sm.transition(OrderStatus.CANCEL_REQUESTED);
    assert.equal(sm.status, OrderStatus.CANCEL_REQUESTED);

    sm.transition(OrderStatus.CANCELLED);
    assert.equal(sm.status, OrderStatus.CANCELLED);
    assert.equal(sm.isTerminal, true);
  });

  it('calls onTransition callback on state change', () => {
    const transitions: [string, string][] = [];
    const sm = new OrderStateMachine(OrderStatus.CREATED, (from, to) => {
      transitions.push([from, to]);
    });

    sm.transition(OrderStatus.SUBMITTED);
    sm.transition(OrderStatus.PENDING);

    assert.deepEqual(transitions, [
      [OrderStatus.CREATED, OrderStatus.SUBMITTED],
      [OrderStatus.SUBMITTED, OrderStatus.PENDING]
    ]);
  });
});
